src/client/visualizations/grid/grid.tsx

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2017-2018 Allegro.pl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as d3 from "d3";
+import { Dataset, Datum, Expression, PseudoDatum } from "plywood";
+import * as React from "react";
+import { Essence } from "../../../common/models/essence/essence";
+import { Timekeeper } from "../../../common/models/timekeeper/timekeeper";
+import { Direction, ResizeHandle } from "../../components/resize-handle/resize-handle";
+import { Scroller, ScrollerLayout } from "../../components/scroller/scroller";
+import { selectFirstSplitDatums } from "../../utils/dataset/selectors/selectors";
+import { BaseVisualization, BaseVisualizationState } from "../base-visualization/base-visualization";
+import { FlattenedSplits } from "../table/body/splits/flattened-splits";
+import { MeasuresHeader } from "../table/header/measures/measures-header";
+import { SplitColumnsHeader } from "../table/header/splits/split-columns";
+import { HEADER_HEIGHT, ROW_HEIGHT, SPACE_LEFT } from "../table/table";
+import { measureColumnsCount } from "../table/utils/measure-columns-count";
+import { visibleIndexRange } from "../table/utils/visible-index-range";
+import makeQuery from "./make-query";
+import { MeasureRows } from "./measure-rows";
+
+interface GridState extends BaseVisualizationState {
+  segmentWidth: number;
+}
+
+const MIN_DIMENSION_WIDTH = 100;
+const SEGMENT_WIDTH = 300;
+const MEASURE_WIDTH = 130;
+const SPACE_RIGHT = 10;
+
+export class Grid extends BaseVisualization<GridState> {
+  protected innerGridRef = React.createRef<HTMLDivElement>();
+
+  protected getQuery(essence: Essence, timekeeper: Timekeeper): Expression {
+    return makeQuery(essence, timekeeper);
+  }
+
+  getDefaultState(): GridState {
+    return {
+      segmentWidth: SEGMENT_WIDTH,
+      ...super.getDefaultState()
+    };
+  }
+
+  setScroll = (scrollTop: number, scrollLeft: number) => this.setState({ scrollLeft, scrollTop });
+
+  setSegmentWidth = (segmentWidth: number) => this.setState({ segmentWidth });
+
+  private getIdealColumnWidth(): number {
```

Comment:
How to test calculation if method is private? I would extract layout math form the component class.

Reply:
Probably you're right. As always, these functions mostly connect multiple variables from object state so tests will get messy. But extraction is good first step!